File: multi-gammon-core/src/main/java/org/akoshterek/backgammon/agent/fa/FunctionApproximator.java
package org.akoshterek.backgammon.agent.fa;

import org.akoshterek.backgammon.eval.Reward;

/**
 * @author Alex
 *         date 02.08.2015.
 */
public interface FunctionApproximator {
    Reward calculateReward(double[] input);
    void setReward(double[] input, Reward reward);

    default void updateAddToReward(double[] input, Reward deltaReward)
    {
        Reward currentReward = calculateReward(input);
        setReward(input, Reward.plus(currentReward, deltaReward).clamp());
    }
}

File: multi-gammon-core/src/main/java/org/akoshterek/backgammon/agent/fa/FunctionApproximator.scala
package org.akoshterek.backgammon.agent.fa

import org.akoshterek.backgammon.eval.Reward

/**
  * @author Alex
  *         date 02.08.2015.
  */
trait FunctionApproximator {
    def calculateReward(input: Array[Float]): Reward

    def setReward(input: Array[Float], reward: Reward): Unit

    def updateAddToReward(input: Array[Float], deltaReward: Reward): Unit = {
        val currentReward: Reward = calculateReward(input)
        setReward(input, (currentReward + deltaReward).clamp())
    }
}